lib/file.c

```diff
@@ -380,126 +387,222 @@ static CURLcode file_upload(struct Curl_easy *data)
   return result;
 }
 
-/*
- * file_do() is the protocol-specific function for the do-phase, separated
- * from the connect-phase above. Other protocols merely setup the transfer in
- * the do-phase, to have it done in the main transfer loop but since some
- * platforms we support don't allow select()ing etc on file handles (as
- * opposed to sockets) we instead perform the whole do-operation in this
- * function.
- */
-static CURLcode file_do(struct Curl_easy *data, bool *done)
+/* Convert a permission bitmask to a string (e.g. 0777 = -rwxrwxrwx) */
+static void fmt_mode(char *str, mode_t mode)
 {
-  /* This implementation ignores the host name in conformance with
-     RFC 1738. Only local files (reachable via the standard file system)
-     are supported. This means that files on remotely mounted directories
-     (via NFS, Samba, NT sharing) can be accessed through a file:// URL
-  */
-  CURLcode result = CURLE_OK;
-  struct_stat statbuf; /* struct_stat instead of struct stat just to allow the
-                          Windows version to have a different struct without
-                          having to redefine the simple word 'stat' */
-  curl_off_t expected_size = -1;
-  bool size_known;
-  bool fstated = FALSE;
-  char *buf = data->state.buffer;
-  curl_off_t bytecount = 0;
-  int fd;
-  struct FILEPROTO *file;
+  memset(str, '\0', MODE_LEN);
+
+  if(S_ISBLK(mode))
+    str[0] = 'b';
+  else if(S_ISCHR(mode))
+    str[0] = 'c';
+  else if(S_ISDIR(mode))
+    str[0] = 'd';
+  else if(S_ISFIFO(mode))
+    str[0] = 'p';
+  else if(S_ISLNK(mode))
+    str[0] = 'l';
+  else if(S_ISREG(mode))
+    str[0] = '-';
+  else if(S_ISSOCK(mode))
+    str[0] = 's';
+  else
+    str[0] = '?';
+
+  str[1] = mode & S_IRUSR ? 'r' : '-';
+  str[2] = mode & S_IWUSR ? 'w' : '-';
+  str[3] = mode & S_IXUSR ? 'x' : '-';
+  str[4] = mode & S_IRGRP ? 'r' : '-';
+  str[5] = mode & S_IWGRP ? 'w' : '-';
+  str[6] = mode & S_IXGRP ? 'x' : '-';
+  str[7] = mode & S_IROTH ? 'r' : '-';
+  str[8] = mode & S_IWOTH ? 'w' : '-';
+  str[9] = mode & S_IXOTH ? 'x' : '-';
+}
 
-  *done = TRUE; /* unconditionally */
+static CURLcode fmt_mtime(char *str, time_t mtime)
+{
+  struct tm tm, now;
+  CURLcode result;
 
-  Curl_pgrsStartNow(data);
+  result = Curl_gmtime(mtime, &tm);
+  if(result)
+    return result;
+  result = Curl_gmtime(time(NULL), &now);
+  if(result)
+    return result;
 
-  if(data->set.upload)
-    return file_upload(data);
+  if(tm.tm_year == now.tm_year)
+    strftime(str, MTIME_LEN, "%b %d %H:%M", &tm);
+  else
+    strftime(str, MTIME_LEN, "%b %d %Y", &tm);
```

Comment:
Would it easier to parse with the year always included? It would also allow parsing without knowing in which year the request was returned.

Unless there is a convention or standard to follow there, parsing may be even easier if using an ISO timestamp here, e.g.: `%Y-%m-%dT%H:%M:%SZ`. With the bonus of having seconds included.

Reply:
I too would prefer a stable syntax, independent of time passing.

> Unless there is a convention or standard to follow

I guess @emilengler tried to make this look like an FTP listing so the convention would then be the format common FTP servers (ls implementations?) show.

Reply:
Done e308e3d8dfccf3ddf5eaa46c5c34dd3be0539dbc